sources/zarr/large_image_source_zarr/__init__.py

```diff
@@ -438,6 +492,203 @@ def getTile(self, x, y, z, pilImageAllowed=False, numpyAllowed=False, **kwargs):
         return self._outputTile(tile, TILE_FORMAT_NUMPY, x, y, z,
                                 pilImageAllowed, numpyAllowed, **kwargs)
 
+    def addTile(self, tile, x=0, y=0, mask=None, axes=None, **kwargs):
+        """
+        Add a numpy or image tile to the image, expanding the image as needed
+        to accommodate it.  Note that x and y can be negative.  If so, the
+        output image (and internal memory access of the image) will act as if
+        the 0, 0 point is the most negative position.  Cropping is applied
+        after this offset.
+
+        :param tile: a numpy array, PIL Image, or a binary string
+            with an image.  The numpy array can have 2 or 3 dimensions.
+        :param x: location in destination for upper-left corner.
+        :param y: location in destination for upper-left corner.
+        :param mask: a 2-d numpy array (or 3-d if the last dimension is 1).
+            If specified, areas where the mask is false will not be altered.
+        :param axes: a string or list of strings specifying the names of axes
+            in the same order as the tile dimensions
+        :param kwargs: start locations for any additional axes
+        """
+        # TODO: improve band bookkeeping
+
+        self._checkEditable()
+        placement = {
+            'x': x,
+            'y': y,
+            **kwargs,
+        }
+        if axes is None:
+            if len(tile.shape) == 2:
+                axes = 'yx'
+            elif len(tile.shape) == 3:
+                axes = 'yxs'
+            else:
+                axes = ''
+        if isinstance(axes, str):
+            axes = axes.lower()
+        elif isinstance(axes, list):
+            axes = [x.lower() for x in axes]
+        else:
+            err = 'Invalid type for axes. Must be str or list[str].'
+            raise ValueError(err)
+        self._axes = {k: i for i, k in enumerate(axes)}
+
+        tile, mode = _imageToNumpy(tile)
+        while len(tile.shape) < len(axes):
+            tile = np.expand_dims(tile, axis=0)
+
+        new_dims = {
+            a: max(
+                self._dims.get(a, 0),
+                placement.get(a, 0) + tile.shape[i],
+            )
+            for a, i in self._axes.items()
+        }
+        self._dims = new_dims
+        self._dtype = tile.dtype
+        self._bandCount = new_dims.get(axes[-1])  # last axis is assumed to be bands
+        self.sizeX = new_dims.get('x')
+        self.sizeY = new_dims.get('y')
+        self._framecount = np.prod([
+            length
+            for axis, length in new_dims.items()
+            if axis in axes[:-3]
+        ])
+        self._levels = None
+        self.levels = int(max(1, math.ceil(math.log(max(
+            self.sizeX / self.tileWidth, self.sizeY / self.tileHeight)) / math.log(2)) + 1))
+
+        if not mask:
+            mask = np.full(tile.shape, True)
+        full_mask = np.full(tuple(new_dims.values()), False)
```

Comment:
There really is no need to create this full mask.  It will not scale.  I want to build an image that won't fit in memory; this mask will be the size of the output and therefore won't fit in memory either.  I also wouldn't create a mask if none is specified.

Instead, `np.place(data, full_mask, tile)` can probably be 
```
if mask is not None:
    np.place(root[mask_placement_slices], mask, tile)
else:
   root[mask_placement_slices] = tile
```

Reply:
Changed in [cd0c715](https://github.com/girder/large_image/pull/1446/commits/cd0c715d61812c4e3e2eefddcf92dd659a65fe89). I had to do a slight variation because `np.place` won't edit the zarr data in place, so I used a placeholder variable "data" which contains the sliced section of the zarr array.

Reply:
Maybe the `np.where` idiom would be clearer so you don't need the `data` placeholder variable:
```
if mask:
    root[mask_placement_slices] = np.where(mask, tile, root[mask_placement_slices])
else:
    root[mask_placement_slices] = tile
```

Reply:
Changed in [41ea6b5](https://github.com/girder/large_image/pull/1446/commits/41ea6b5e7938dc867122ad6d3da5836b58a47aed)